docs/proposals/04-etcd-member-custom-resource.md

```diff
@@ -0,0 +1,545 @@
+
+* Today [leases](https://kubernetes.io/docs/concepts/architecture/leases/) are misused to share member-specific information with etcd-druid. Their usage to share member state [leader, follower, learner], member-id, snapshot revisions etc should be removed.
+
+### Non-Goals
+
+* Auto-recovery from quorum loss or cluster-split due to network partitioning.
+* Auto-recovery of an etcd-member due to volume mismatch.
+
+## Proposal
+
+This proposal introduces a new custom resource `EtcdMember`, and in the following sections describes different sets of information that should be captured as part of the new resource.
+
+### Etcd Member Metadata
+
+Every etcd-member has a unique `memberID` and it is part of an etcd cluster which has a unique `clusterID`. In a well-formed etcd cluster every member must have the same `clusterID`. Publishing this information to druid helps in identifying issues when one or more etcd-members form their own individual clusters, thus resulting in multiple clusters where only one was expected. Issues [Issue#361](https://github.com/gardener/etcd-druid/issues/361) and [Issue#419](https://github.com/gardener/etcd-druid/issues/419) are some such occurrences.
+
+Today, this information is published by using a member [lease](https://kubernetes.io/docs/concepts/architecture/leases/). Both these fields are populated in the leases' `Spec.HolderIdentity` by the backup-sidecar container. 
+
+The authors propose to publish member metadata information in `EtcdMember` resource.
+
+```yaml
+id: <etcd-member id>
+clusterID: <etcd cluster id>
+```
+
+> **NOTE:** Druid would not do any auto-recovery when it finds out that there are more than one clusters being formed. Instead this information today will be used for diagnostic and alerting.
+
+### Etcd Member State Transitions
+
+Each etcd-member goes through different `States` during its lifetime. `State` is a derived high-level summary of where an etcd-member is in its lifecycle. A `SubState` gives additional information about the state. This proposal extends the concept of states with the notion of a `SubState`, since `State` indicates a top-level state of an `EtcdMember` resource, which can have one or more `SubState`s.
+
+While `State` is sufficient for many human operators, the notion of a `SubState` provides operators with an insight about the discrete stage of an etcd-member in its lifecycle. For example, consider a top-level `State: Starting`, which indicates that an etcd-member is starting. `Starting` is meant to be a transient state for an etcd-member. If an etcd-member remains in this `State` longer than expected, then an operator would require additional insight, which the authors propose to provide via `SubState` (in this case, the possible `SubStates` could be `PendingLearner` and `Learner`, which are detailed in the following sections).
```

Comment:
Is the `SubState` meant for operators only? If yes, how about providing a field `description` instead?

Reply:
This is going to be part of the state machine that we are going to create. It is not only meant for developers but also will be consumed by druid.

Reply:
Got it. While this is valid I think that `state`  or especially `substate` is not very intuitive and consistent compared to other `status` subresources of Gardener or Kubernetes objects. Maybe you find some field names that are closer to our ecosystem?

Reply:
we were divided between using `phase` and `subPhase` vs `state` and `subState` and we chose the latter.
Pods have a top level phase which essentially is derived by looking at conditions.  So we could align to that terminology as well.

Reply:
You mean to have `phase` for top level and `conditions` (alternatively `lastOperation`) to determine more information about the phase?

Reply:
We evaluated using conditions but it was getting too complicated to represent state transitions and therefore we did not go down that path and instead stuck to using simple, easily interpreted state transitions. So for `EtcdMember` resource we do not envision any conditions as of now. In future if there is a need we could introduce it.